plutus-contract/src/Plutus/Contract/Test/StateModel.hs

```diff
@@ -0,0 +1,176 @@
+-- This is a simple state modelling library for use with Haskell
+-- QuickCheck.
+
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE StandaloneDeriving    #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module Plutus.Contract.Test.StateModel(
+    StateModel(..)
+  , Any(..)
+  , Step(..)
+  , LookUp, Var(..) -- we export the constructors so that users can construct test cases
+  , Actions(..)
+  , stateAfter
+  , runScript
+  , runScriptInState
+) where
+
+import           Data.Typeable
+
+import           Test.QuickCheck         as QC
+import           Test.QuickCheck.Monadic
+
+class (forall a. Show (Action state a),
+       Monad (ActionMonad state),
+       Show state,
+       Typeable state) =>
+        StateModel state where
+  data Action state a
+  type ActionMonad state :: * -> *
+  actionName      :: Action state a -> String
+  actionName = head . words . show
+  arbitraryAction :: state -> Gen (Any (Action state))
+  shrinkAction    :: (Typeable a, Show a) => state -> Action state a -> [Any (Action state)]
+  shrinkAction _ _ = []
+  initialState    :: state
+  nextState       :: state -> Action state a -> Var a -> state
+  nextState s _ _ = s
+  precondition    :: state -> Action state a -> Bool
+  precondition _ _ = True
+  perform         :: state -> Action state a -> LookUp -> ActionMonad state a
+  perform _ _ _ = return undefined
+  postcondition   :: state -> Action state a -> LookUp -> a -> Bool
+  postcondition _ _ _ _ = True
+  monitoring      :: (state,state) -> Action state a -> LookUp -> a -> Property -> Property
+  monitoring _ _ _ _ = id
+
+type LookUp = forall a. Typeable a => Var a -> a
```

Comment:
So this *can* actually fail, we just die with error. Why not have this return a `Maybe`, and then it's up to the implementor of `perform` to decide what to do there (could be "die with error", could be something else in `ActionMonad`).

Reply:
If this fails your model is broken and you want to "die with error". I don't think there's any scenario where you would want to handle it in some other way.